server.go

```diff
@@ -3520,63 +3533,75 @@ func (s *server) peerTerminationWatcher(p *peer.Brontide, ready chan struct{}) {
-					"to inbound peer %v without "+
+				srvrLog.Debugf("Ignoring reconnection "+
+					"attempt to inbound peer %v without "+
 					"advertised address", p)
 				return
 
 			// We came across an error retrieving an advertised
 			// address, log it, and fall back to the existing peer
 			// address.
 			default:
-				srvrLog.Errorf("Unable to retrieve advertised "+
-					"address for node %x: %v", p.PubKey(),
-					err)
+				srvrLog.Errorf("Unable to retrieve "+
+					"advertised address for node %x: "+
+					"%v", p.PubKey(), err)
 			}
-		}
 
-		// Otherwise, we'll launch a new connection request in order to
-		// attempt to maintain a persistent connection with this peer.
-		connReq := &connmgr.ConnReq{
-			Addr:      p.NetAddress(),
-			Permanent: true,
-		}
-		s.persistentConnReqs[pubStr] = append(
-			s.persistentConnReqs[pubStr], connReq)
+			s.mu.Lock()
 
-		// Record the computed backoff in the backoff map.
-		backoff := s.nextPeerBackoff(pubStr, p.StartTime())
-		s.persistentPeersBackoff[pubStr] = backoff
+			// Create a new connection request in order to attempt
+			// to maintain a persistent connection with this peer.
+			connReq := &connmgr.ConnReq{
+				Addr:      p.NetAddress(),
+				Permanent: true,
+			}
+			s.persistentConnReqs[pubStr] = append(
+				s.persistentConnReqs[pubStr], connReq)
 
-		// Initialize a retry canceller for this peer if one does not
-		// exist.
-		cancelChan, ok := s.persistentRetryCancels[pubStr]
-		if !ok {
-			cancelChan = make(chan struct{})
-			s.persistentRetryCancels[pubStr] = cancelChan
-		}
+			// Record the computed backoff in the backoff map.
+			backoff := s.nextPeerBackoff(pubStr, p.StartTime())
+			s.persistentPeersBackoff[pubStr] = backoff
+
+			// Initialize a retry canceller for this peer if one
+			// does not exist.
+			cancelChan, ok := s.persistentRetryCancels[pubStr]
+			if !ok {
+				cancelChan = make(chan struct{})
+				s.persistentRetryCancels[pubStr] = cancelChan
+			}
 
-		// We choose not to wait group this go routine since the Connect
-		// call can stall for arbitrarily long if we shutdown while an
-		// outbound connection attempt is being made.
-		go func() {
-			srvrLog.Debugf("Scheduling connection re-establishment to "+
-				"persistent peer %v in %s", p, backoff)
+			s.mu.Unlock()
+
+			srvrLog.Debugf("Scheduling connection "+
+				"re-establishment to persistent peer %v in "+
+				"%s", p, backoff)
 
 			select {
 			case <-time.After(backoff):
+			// If cancelChan is closed then the peer has been
+			// connected.
 			case <-cancelChan:
 				return
 			case <-s.quit:
 				return
 			}
 
-			srvrLog.Debugf("Attempting to re-establish persistent "+
-				"connection to peer %v", p)
+			// If we have already created a connection request
+			// with the same address, then there is no need to
+			// create a new one.
+			if prevConnReqAddr != nil && connReq.Addr.String() ==
+				prevConnReqAddr.String() {
+
+				continue
```

Comment:
> But not sure if there is a better way? would maybe be fine if the backoff period was reset once the connection was reestablished but at the moment it doesnt do that)

Yeah the purpose of this back off is to ensure that if we connect, then the peer disconnects, we don't try to reconnect to it in a tight loop. The result of this new `for` loop is that we'll do many iterations here until we hit the back off upper limit then start to slow down some.

Reply:
I think one way to address this would be to layer another back off (local to this method) to clamp down on the cycle speed of this new loop. In practice this should serve to ensure that we don't burn the greater back off (which should only be increased each time we disconnect from the peer, and not each time we try to refresh their address set).